src/oauth-provider.ts

```diff
@@ -1650,43 +1679,82 @@ class OAuthProviderImpl {
+      newRefreshTokenId = await generateTokenId(newRefreshToken);
+      newRefreshTokenWrappedKey = await wrapKeyWithToken(newRefreshToken, grantEncryptionKey);
+
+      // Update the grant with the token rotation information
+      // The token which the client used this time becomes the "previous" token, so that the client
+      // can always use the same token again next time. This might technically violate OAuth 2.1's
+      // requirement that refresh tokens be single-use. However, this requirement violates the laws
+      // of distributed systems. It's important that the client can always retry when a transient
+      // failure occurs. Under the strict requirement, if the failure occurred after the server
+      // rotated the token but before the client managed to store the updated token, then the client
+      // no longer has any valid refresh token and has effectively lost its grant. That's bad! So
+      // instead, we don't invalidate the old token until the client successfully uses a newer token.
+      // This provides most of the security benefits (tokens still rotate naturally) but without
+      // being inherently unreliable.
+      grantData.previousRefreshTokenId = providedTokenHash;
+      grantData.previousRefreshTokenWrappedKey = wrappedKeyToUse;
+
+      // The newly-generated token becomes the new "current" token
+      grantData.refreshTokenId = newRefreshTokenId;
+      grantData.refreshTokenWrappedKey = newRefreshTokenWrappedKey;
+    } else {
+      // Final refresh: no token rotation. The client keeps using their existing refresh token(s)
+      // for retries. Unlike normal refresh, we don't move tokens between current/previous slots -
+      // both tokens remain valid in their current positions. This makes the operation idempotent:
+      // whether the client successfully receives the response or not, the grant's token state
+      // remains unchanged, ensuring reliable retries in distributed systems.
+      //
+      // We update the wrapped key for whichever token was used (current or previous) with the
+      // key that successfully authenticated this request. This ensures retries work correctly,
+      // especially when props change (which changes the grant's encryption key).
+      if (isCurrentToken) {
+        grantData.refreshTokenWrappedKey = wrappedKeyToUse;
+      } else {
+        grantData.previousRefreshTokenWrappedKey = wrappedKeyToUse;
+      }
+    }
 
     // Update refresh token expiration if a new TTL was provided by callback, otherwise keep existing expiration
     if (refreshTokenTTLUpdated) {
-      // If TTL is 0 or null, remove expiration; otherwise set new expiration
-      grantData.refreshTokenExpiresAt = refreshTokenTTL ? now + refreshTokenTTL : undefined;
+      if (isFinalRefresh) {
+        // For final refresh, we don't update expiration - the current token keeps its existing expiration
+        // This allows the client to retry this final refresh if needed
```

Comment:
Is this behaviour a fair expectation? From user pov

Reply:
No, I don't think it is.

`refreshTokenTTL = 0` is supposed to mean "don't issue a refresh token". But here we _already_ issued one, but on a refresh operation, the callback returned this. I guess the most sensible thing to do would be to delete the existing refresh token.

The grant record needs to continue existing, though, until all access tokens are gone. So I guess you'd want to set its TTL in KV to match the access token.

This all seems like a lot of work to support something nobody asked for, though. Maybe we just shouldn't support changing the TTL after the grant is created?